update performance tips

KristofferC opened the pull request:
A lot of the stuff here was terribly dated, and the examples were no good anymore due to new optimizations like union splitting.

I've made some comments inline to give some reasons for the changes.

https://github.com/JuliaLang/julia/pull/27033/files?w=1

Fixes https://github.com/JuliaLang/julia/issues/27029

KristofferC committed:
update performance tips

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -21,12 +21,18 @@ const DEFAULT_VAL = 0
 
 ```julia
-global x
-y = f(x::Int + 1)
````
This seemed like a silly example, using a type annotation to improve the performance of a single plus. Changed it to instead be something like in a loop.

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -48,79 +54,79 @@ so all the performance issues discussed previously apply.
+A useful tool for measuring performance is the [`@time`](@ref) macro. We here repeat the example
+with the global variable above, but this time with the type annotation removed:
 
-```julia-repl
````
Changed this from a 2-union type instability to global variables since unions are pretty fast now and more significantly do not allocate anymore, which was one of the point of this section.

KristofferC commented on doc/src/manual/performance-tips.md:
```diff
@@ -48,79 +54,79 @@ so all the performance issues discussed previously apply.
+    `*.mem` files to see information about where those allocations occur. See [Memory allocation analysis](@ref).
   * `@code_warntype` generates a representation of your code that can be helpful in finding expressions
     that result in type uncertainty. See [`@code_warntype`](@ref) below.
-  * The [Lint](https://github.com/tonyhffong/Lint.jl)
```
I don't think this is maintained enough to be linked to.

jekbradbury replied:
Is Traceur.jl ready to be linked here?

KristofferC replied:
Added

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -129,23 +135,36 @@ abstract types where possible.
 
-```julia
-a = Real[]    # typeof(a) = Array{Real,1}
-if (f = rand()) < .8
````
Changed this example to show that a `Real[]` can store a bunch of different random stuff to better make the point that it must store things "inefficently". Contrast that with a `Float64[]` below.

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -295,11 +313,11 @@ to resolve the type at run-time. This results in shorter and faster code.
 The same best practices also work for container types:
 
 ```jldoctest containers
-julia> mutable struct MySimpleContainer{A<:AbstractVector}
````
No reason for these to be mutable.

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -355,109 +373,36 @@ foo (generic function with 2 methods)
-end
-```
-
-This works fine for `Vector{T}`, but we'd also have to write explicit versions for `UnitRange{T}`
````
This whole thing seems redundant now since we can just do triangular dispatch. Changed to that instead.

KristofferC commented on doc/src/manual/performance-tips.md:
```diff
@@ -599,8 +546,7 @@ optimize the body of the loop. There are several possible fixes:
 
   * Initialize `x` with `x = 1.0`
   * Declare the type of `x`: `x::Float64 = 1`
-  * Use an explicit conversion: `x = oneunit(T)`
```
In `oneunit(T)`, what is `T` even?

stevengj replied:
Here, it seems like `oneunit(Float64)` would parallel the other examples.

KristofferC commented on doc/src/manual/performance-tips.md:
````diff
@@ -1173,44 +1117,45 @@ The common idiom of using 1:n to index into an AbstractArray is not safe if the
 
 ```julia
-function inner(x, y)
+@noinline function inner(x, y)
````
Have to put `@noinline` here or this gets inlined and optimized away in `timeit` below. See https://github.com/JuliaLang/julia/issues/27029

KristofferC commented on doc/src/manual/performance-tips.md:
```diff
@@ -1394,91 +1366,109 @@ a = rand(Float32,1000) * 1.f-9
 is that non-concrete types are displayed in red; in the above example, such output is shown in
 all-caps.
 
 The top part of the output summarizes the type information for the different variables internal
```
The text here described output from ´@code_warntype` in a way earlier julia version than what we actually showed here. So someone likely updated the output but not the text below.

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -21,12 +21,18 @@ const DEFAULT_VAL = 0
 
-Writing functions is better style. It leads to more reusable code and clarifies what steps are
-being done, and what their inputs and outputs are.
+Passing arguments into functions is better style. It leads to more reusable code and clarifies what the inputs and outputs are.
```
into -> to?

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -48,79 +54,79 @@ so all the performance issues discussed previously apply.
-julia> @time f_improved(1)
-  0.007008 seconds (1.32 k allocations: 63.640 KiB)
-0.5
+julia> @time sum_arg(x)
```
`sum_arg` hasn't been defined anywhere?

KristofferC replied:
Neither was `f_improved`. I think the point before was to not define it to spur the readers interest but it is a bit more obvious what the problem is now since it has been discussed earlier.

nalimilan replied:
But why not define `sum_arg(x)` just like we defined `sum_global` above?

KristofferC replied:
Did that

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -249,11 +268,11 @@ julia> typeof(m.a)
 
-The fact that the type of `m.a` is known from `m`'s type--coupled with the fact that its type
-cannot change mid-function--allows the compiler to generate highly-optimized code for objects
+The fact that the type of `m.a` is known from `m`'s type—coupled with the fact that its type
```
While you're at it, add spaces around the dashes?

stevengj replied:
It varies.   e.g. [this source](http://www.thepunctuationguide.com/em-dash.html) says that in AP style you have a space before and after em dashes, whereas in [Chicago style you have no spaces](https://apvschicago.com/2011/05/em-dashes-and-ellipses-closed-or-spaced.html).

(In a monospaced font, I find the em dash much more readable if it is set off with spaces, because the monospaced em is too narrow.  The manual is displayed in a proportional font, of course, but we edit it in monospace.)

Either way seems fine to me, as long as we are consistent.

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -355,109 +373,36 @@ foo (generic function with 2 methods)
 
 However, there are cases where you may need to declare different versions of the outer function
-for different element types of `a`. You could do it like this:
+for different element types or types of the `AbstractVector` of the field `a` in `MySimpleContainer`. You could do it like this:
```
Break at 92 chars.

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -355,109 +373,36 @@ foo (generic function with 2 methods)
 myfunc (generic function with 1 method)
 
-julia> function myfunc(c::MyContainer{<:AbstractFloat})
+julia> function myfunc(c::MySimpleContainer{<:AbstractArray{T}}) where T <: AbstractFloat
```
Why not just `<:AbstractArray{<:AbstractFloat}}`?

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -599,8 +546,7 @@ optimize the body of the loop. There are several possible fixes:
   * Declare the type of `x`: `x::Float64 = 1`
-  * Use an explicit conversion: `x = oneunit(T)`
-  * Initialize with the first loop iteration, to `x = 1/bar()`, then loop `for i = 2:10`
+  * Initialize with the first loop iteration, to `x = 1 / rand()`, then loop `for i = 2:10`
```
`rand` seems even more confusing than `bar` here. Though this section clearly needs to be developed more.

KristofferC replied:
With `bar` it wasn't even the case that the result had to be a float so that's why I changed it.

nalimilan replied:
OK. That section isn't completely well constructed anyway...

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -609,22 +555,14 @@ to perform a core computation. Where possible, it is a good idea to put these co
            return a
-       end
-strange_twos (generic function with 1 method)
+       end;
```
Maybe that's just me, but I find it weird to end functions with `;` just to avoid printing it.

KristofferC replied:
Maybe, but I rather save the vertical space.

nalimilan commented on doc/src/manual/performance-tips.md:
```diff
@@ -1329,6 +1274,20 @@ is much faster to evaluate. Of course, both the actual optimization that is appl
 as well as the resulting speedup depend very much on the hardware. You can examine the change
 in generated code by using Julia's [`code_native`](@ref) function.
 
+Note that `@fastmath` also assumes that NaNs will not occur during the computation, which can lead to surprising behavior:
```
`` `NaN`s``

KristofferC committed:
address review, add doctests, update to new code_warntype output

KristofferC commented:
- Updated from review comments.
- Made quite a few things doctests instead of `julia-repl`.
- Updated to new `code_warntype` output.

Will merge in not too long time unless more comments appear.

KristofferC committed:
fix doctest

KristofferC committed:
fine

jebej commented:
I was trying to understand why inlining made a difference given that no `SIMD` condition in the tips were being broken (it was on the inner loop etc...):

It's because the result is not used and so the first loop gets optimized away. It might be clearer to explain that rather than just saying that the compiler is being clever.

A nice way to fix it, would be to actually print the result, which would also show the difference between the two methods due to operation reordering:

```julia
function timeit2(n, reps)
    x = rand(Float32,n)
    y = rand(Float32,n)
    s = zero(Float64)
    time = @elapsed for j in 1:reps
        s+=inner(x,y)
    end
    println("Non-SIMD, res = $(round(Int,s)), GFlop/sec = ",2.0*n*reps/time*1E-9)
    s = zero(Float64)
    time = @elapsed for j in 1:reps
        s+=innersimd(x,y)
    end
    println("SIMD,     res = $(round(Int,s)), GFlop/sec = ",2.0*n*reps/time*1E-9)
end
```

with result (on my terrible computer):

```julia
julia> timeit2(10^6,100)
Non-SIMD, res = 25013767, GFlop/sec = 2.1076901075880956
SIMD,     res = 25017553, GFlop/sec = 4.5190580086847705
```

KristofferC commented:
Sounds like an excellent improvement, go for it!
